File: 1043-partition-array-for-maximum-sum/1043-partition-array-for-maximum-sum.cpp
class Solution {
public:
    int solve(vector<int>& arr, int ind, int k, int n,vector<int>& dp)
    {
        if(ind>=n) return 0;

        if(dp[ind]!=-1) return dp[ind];

        int max_ele = INT_MIN;
        int sum = 0;
        for(int i=ind; i<min(ind+k,n); i++)
        {
            max_ele = max(max_ele,arr[i]);
            sum = max(sum , solve(arr,i+1,k,n,dp)+(i-ind+1)*max_ele);
        }
        return dp[ind] = sum;
    }
    int maxSumAfterPartitioning(vector<int>& arr, int k) {
        int n = arr.size();
        vector<int> dp(n+1,-1);
        return solve(arr,0,k,n,dp);
    }
};

File: 1043-partition-array-for-maximum-sum/1043-partition-array-for-maximum-sum.py
class Solution:
    def maxSumAfterPartitioning(self, arr: List[int], k: int) -> int:
        def helper(i,arr,n,k,dp):
            if(i==n):
                return 0
        
            if(dp[i]!=-1) :
                return dp[i]
        
            maxi=-10**9
            maxians=-10**9
            leng=0
            for j in range(i,min(n,i+k)):
                leng+=1
                maxi=max(maxi,arr[j])
                cost=(maxi*leng) + helper(j+1,arr,n,k,dp)
                maxians=max(maxians,cost) 
        
            dp[i]=maxians 
            return dp[i]
        dp=[-1]*(len(arr))
        return helper(0,arr,len(arr),k,dp)